Sources/AsyncAlgorithms/AsyncShareSequence.swift

```diff
@@ -0,0 +1,693 @@
+            return (side, nil, true, nil)
+          } else {
+            defer {
+              if cancelRequested {
+                state.iteratingTask = .cancelled
+              }
+            }
+            return (side, nil, false, cancelRequested ? state.iteratingTask : nil)
+          }
+        }
+      }
+      if let continuation {
+        continuation.resume(returning: cancelled)
+      }
+      if let side {
+        side.continuation?.resume(returning: .success(nil))
+      }
+      if let iteratingTaskToCancel {
+        iteratingTaskToCancel.cancel()
+      }
+    }
+    
+    func iterate() async -> Bool {
+      if let limit {
+        let cancelled = await withUnsafeContinuation { (continuation: UnsafeContinuation<Bool, Never>) in
+          let (resume, cancelled) = state.withLock { state -> (UnsafeContinuation<Bool, Never>?, Bool) in
+            if state.buffer.count >= limit {
+              state.limit = continuation
+              if case .cancelled = state.iteratingTask {
+                return (nil, true)
+              } else {
+                return (nil, false)
+              }
+            } else {
+              assert(state.limit == nil)
+              if case .cancelled = state.iteratingTask {
+                return (continuation, true)
+              } else {
+                return (continuation, false)
+              }
+            }
+          }
+          if let resume {
+            resume.resume(returning: cancelled)
+          }
+        }
+        if cancelled {
+          return false
+        }
+      }
+      
+      // await a demand
+      await withUnsafeContinuation { (continuation: UnsafeContinuation<Void, Never>) in
+        let hasPendingDemand = state.withLock { state in
+          for (_, side) in state.sides {
+            if side.continuation != nil {
+              return true
+            }
+          }
+          state.demand = continuation
+          return false
+        }
+        if hasPendingDemand {
+          continuation.resume()
+        }
+      }
+      return state.withLock { state in
+        switch state.iteratingTask {
+        case .cancelled:
+          return false
+        default:
+          return true
+        }
+      }
+    }
+    
+    func cancel(id: Int) {
+      unregisterSide(id) // doubly unregistering is idempotent but has a side effect of emitting nil if present
+    }
+    
+    struct Resumption {
+      let continuation: UnsafeContinuation<Result<Element?, Failure>, Never>
+      let result: Result<Element?, Failure>
+      
+      func resume() {
+        continuation.resume(returning: result)
+      }
+    }
+    
+    func emit(_ result: Result<Element?, Failure>) {
+      let (resumptions, limitContinuation, demandContinuation, cancelled) = state.withLock { state -> ([Resumption], UnsafeContinuation<Bool, Never>?, UnsafeContinuation<Void, Never>?, Bool) in
+        var resumptions = [Resumption]()
```

Comment:
Instead of allocating a new array every time you might be better off re-using one